vyper/semantics/validation/module.py

```diff
@@ -167,15 +168,34 @@ def visit_AnnAssign(self, node):
                     # we need this when builing the public getter
                     node._metadata["func_type"] = ContractFunction.from_AnnAssign(node)
 
+                elif call_name == "immutable":
+                    # declaring an immutable variable
+                    is_immutable = True
+
+                    # mutability is checked automatically preventing assignment
+                    # outside of the constructor, here we just check a value is assigned,
+                    # not necessarily where
+                    assignments = self.ast.get_descendants(
+                        vy_ast.Assign, filters={"target.id": node.target.id}
+                    )
+                    if not assignments:
+                        raise SyntaxException(
+                            "Immutable definition requires an assignment in the constructor",
+                            node.node_source_code,
+                            node.lineno,
+                            node.col_offset,
+                        )
+
                 # remove the outer call node, to handle cases such as `public(map(..))`
                 annotation = annotation.args[0]
 
+        data_loc = DataLocation.UNSET if is_immutable else DataLocation.STORAGE
```

Comment:
This is a bit confusing. Why not
```suggestion
        data_loc = DataLocation.CODE if is_immutable else DataLocation.STORAGE
```